rust/src/c/librarydb.rs

```diff
@@ -0,0 +1,231 @@
+//! This module contains the C interface for [`stracciatella::librarydb`].
+//!
+//! [`stracciatella::librarydb`]: ../../librarydb/index.html
+
+use std::io::{self, Read, Seek, SeekFrom};
+
+use crate::c::common::*;
+use crate::librarydb::{LibraryDB, LibraryFile};
+
+/// Constructor.
+/// The caller is responsible for the memory of the library database.
+#[no_mangle]
+pub extern "C" fn LibraryDB_New() -> *mut LibraryDB {
+    into_ptr(LibraryDB::new())
+}
+
+/// Destructor.
+/// The caller is no longer responsible for the memory of the library database.
+#[no_mangle]
+pub extern "C" fn LibraryDB_Delete(ldb: *mut LibraryDB) {
+    let _drop_me = from_ptr(ldb);
+}
+
+/// Adds a library to the end of the library database.
+/// Returns true is successful, false otherwise.
+/// Sets the rust error.
+#[no_mangle]
+pub extern "C" fn LibraryDB_AddLibrary(
+    ldb: *mut LibraryDB,
+    data_dir: *const c_char,
+    library: *const c_char,
+) -> bool {
+    let ldb = unsafe_mut(ldb);
+    let data_dir = path_from_c_str_or_panic(unsafe_c_str(data_dir));
+    let library = path_from_c_str_or_panic(unsafe_c_str(library));
+    match ldb.add_library(data_dir, library) {
+        Err(err) => {
+            remember_rust_error(format!("{:?}", err));
+            false
+        }
+        Ok(_) => {
+            forget_rust_error();
+            true
+        }
+    }
+}
+
+/// Opens a library database file for reading.
+/// Returns the file on success and null on error.
+/// The caller is responsible for the library file memory.
+/// Sets the rust error.
+#[no_mangle]
+pub extern "C" fn LibraryFile_Open(ldb: *mut LibraryDB, path: *const c_char) -> *mut LibraryFile {
+    let ldb = unsafe_mut(ldb);
+    let path = str_from_c_str_or_panic(unsafe_c_str(path));
+    match ldb.open_file(&path) {
+        Err(err) => {
+            remember_rust_error(format!("{:?}", err));
+            std::ptr::null_mut()
+        }
+        Ok(file) => {
+            forget_rust_error();
+            into_ptr(file)
+        }
+    }
+}
+
+/// Closes a library database file.
+/// The caller is no longer responsible for the library file memory.
+#[no_mangle]
+pub extern "C" fn LibraryFile_Close(file: *mut LibraryFile) {
+    let _drop_me = from_ptr(file);
+}
+
+/// Seeks a library database file.
+/// `from` expects FileSeekMode values (see src/sgp/SGPFile.h)
+/// Sets the rust error.
+#[no_mangle]
+pub extern "C" fn LibraryFile_Seek(file: *mut LibraryFile, distance: i64, from: c_int) -> bool {
```

Comment:
Because it's expecting FileSeekMode values (not defined in rust) and C/C++ treats it as an int (not a class enum). It will probably be changed to an enum later, when more filesystem stuff gets rustified.

Reply:
Fine for me.